### ISSUE TITLE
task.t fails with fresh clone and fresh dependencies

### ISSUE THREAD

**Issue by hurricup** [RexOps/Rex#1396]
## Describe the bug

After cloning the repo, and installing dependencies, tests fail in `t/task.t`.

## How to reproduce it

1. install perl-5.31.10 with perlbrew, and activate that version for the shell

    ```bash
    perlbrew install perl-5.31.10
    perlbrew use perl-5.31.10
    ```

2. clone the Rex repo, and switch to that directory

    ```bash
    git clone https://github.com/RexOps/Rex
    cd Rex
    ````
3. install dependencies (_exactly which command(s)?_)
4. run tests (_exactly which command?_)

~~Shortest code example that demonstrates the bug:~~

## Expected behavior

The test suite should pass after cloning the repo, and installing the dependencies.

## Circumstances

 - Rex version: current master (6094895a79b5228e486723c39b4cfff026ebcc31)
 - Perl version: 5.31.10
 - OS running rex: Ubuntu 18 (_exactly which 18?_)
 - OS managed by rex: local 

## Original report

Cloned repo, installed libs, ran tests on Perl 5.31.10

Got failed tests in task.t

```
#   Failed test 'get connection type for ssh'
#   at /home/hurricup/Projects/Rex/t/task.t line 52.
#          got: 'SSH'
#     expected: 'OpenSSH'

#   Failed test 'get connection type for ssh'
#   at /home/hurricup/Projects/Rex/t/task.t line 63.
#          got: 'SSH'
#     expected: 'OpenSSH'
```

**Comment by ferki** [comment]
Hmm, these tests might fail if one or both of `Net::OpenSSH` and `Net::SFTP::Foreign` is not available for some reason. They both should be pulled in as [dependencies on non-Windows systems](https://metacpan.org/source/FERKI/Rex-1.12.1/Makefile.PL#L191-194).

Please check they are both installed, try again, and let me know how that goes!

Since most info expected in a bug report was omitted for some reason, I don't think I can efficiently help much more than the "best guess" above. Therefore I'm going to close this issue for now, but if you still experience this problem, please provide a full bug report either by:

- editing the opening post based on the current bug report template, and then reopening this issue
- opening a new bug report issue with the template filled out

**Comment by ferki** [comment]
For the record, there were some initial Travis CI tests on Ubuntu 18.04 Bionic, and now that is merged into master as part of #1398. Tests still pass, so I'm quite sure it's not related to the Ubuntu version.

On the other hand, it seems that `dzil listdeps` doesn't pick up OS-specific dependencies like `Net::SFTP::Foreign` and `IO::Pty` on Linux (or `Net::SSH2` on Windows). It's because they are injected into `Makefile.PL` through `Dist::Zilla::Plugin::OSPrereqs`, so they are only resolved and installed during `dzil install`.

As suggested by

> Cloned repo, installed libs, ran tests

I _guess_ that the `dzil install` step was not run yet (assuming "installed libs" mean `dzil authordeps --missing | cpanm` followed by `dzil listdeps | cpanm`, and also assuming "ran tests" mean either `prove -l t/` or `dzil test`), therefore optional and OS-specific dependencies were also not present yet.

Good news is that the "install from source" instructions in the README and on the website are still valid. Bad news is that I'm not sure if this behavior can be solved on the Rex side (or even on the Dist::Zilla side).

As a workaround perhaps we can add an explicit test for OS-specific dependencies in order to have a more useful test failure message, or we could improve the conditions of the tests mentioned here.

**Comment by hurricup** [comment]
yep, it wasn't. I'm looking to contribute, not install, so just cloned and ran tests to ensure they are passing on my machine. Failure message was just not really informative. Honestly - haven't installed deps, because haven't got any compilation errors. Will do and check.
I'm not saying it's os dependent, just shared some context when it happened.

**Comment by hurricup** [comment]
Ok. Ran both dependencies installation commands. And still both prove and dzil tests are failing the same way.

**Comment by ferki** [comment]
> I'm looking to contribute

That's appreciated! That just gave me the idea to put some of the info from here into the contributing guide as well.

> Ok. Ran both dependencies installation commands. And still both prove and dzil tests are failing the same way.

Hmm, I just did the full install from scratch and it did work for me. We need to be able to reproduce _your_ exact scenario somehow. So it's probably time to open a full bug report with all the exact details and steps you are trying to run on an empty environment. Please open a new issue, choose the bug report template, and fill out all details.

**Comment by hurricup** [comment]
The reason i've not created a templated bug report is because template looks as a rex bug, not Rex sources.
I've not ran it. Just cloned and ran tests.
If you insist, i can start another issue, but there is no additional information in template i've not provided to you yet. 
Ubuntu 18, perl 5.31.10 from perlbrew, cloned, installed deps, ran tests with prove/dzil. 
I'll try to check in docker and if reproduce - i'll provide you with Dockerfile.

**Comment by ferki** [comment]
> The reason i've not created a templated bug report is because template looks as a Rex bug, not Rex sources.

Thanks, that is helpful feedback! If you feel you need to override the templates, feel free to mention the reason for that right at the start ;)

> I've not ran or installed it. Just cloned repo and ran tests.

As mentioned above, that's not going to work, because the `dzil authordeps` and `dzil listdeps` commands won't install the OS-specific dependencies. One has to run `dzil install` to get those.

> If you insist, i can start another issue, but there is no additional information in template i've not provided to you yet.

Alright, I think I'll reopen and edit this one for you. Even without additional information, the template still has the advantage of presenting the information in an organized format.

> Ubuntu 18, perl 5.31.10 from perlbrew, cloned, installed deps, ran tests with prove/dzil.

The point is we must make sure we run the exact same commands. Please try to be more specific, especially about the exact command used to install the deps, and to run the tests. If I have to keep guessing these details, I will probably end up using different commands.

> I'll try to check in docker and if reproduce - i'll provide you with Dockerfile.

**Comment by hurricup** [comment]
> As mentioned above, that's not going to work, because the dzil authordeps and dzil listdeps commands won't install the OS-specific dependencies. One has to run dzil install to get those.
I'd expected in such case some test failing with appropriate message.

Here is the Dockerfile:

```Dockerfile
from ubuntu:18.04
run apt update
run apt upgrade -y
run apt install -y perlbrew expat libexpat1-dev patch git libssl-dev zlib1g-dev
run echo "source ~/perl5/perlbrew/etc/bashrc" >> ~/.profile
run perlbrew self-upgrade
run perlbrew install --verbose perl-5.31.10 --notest --noman && perlbrew clean
run perlbrew switch perl-5.31.10
run git clone https://github.com/RexOps/Rex.git ~/rex
run cpan install App::cpanminus
run cpanm Dist::Zilla
run mkdir ~/.dzil
run echo "[%User]\nname  = John Doe\nemail = no@no.com" >> ~/.dzil/config.ini
run cd ~/rex/ && dzil authordeps --missing | cpanm
run cd ~/rex/ && dzil listdeps | cpanm
```

Build with: `docker build -t rex_test .`
Run tests with: `docker run -it --rm rex_test bash -c 'cd ~/rex && prove -l t'`

P.S. distzilla is so damn heavy :)

**Comment by ferki** [comment]
> I'd expected in such case some test failing with appropriate message.

Well, to be completely transparent, there are hundreds of reports with success from CPAN Testers, and also each commit is being tested by Travis CI and only merged upon success. Therefore it is always highly interesting if there's a test failure, and especially how the process is different from the official steps. Disclosing the exact reproduction steps right from the start is an important step, so that's why there's a "How to reproduce it" section in the issue template ;)

> Here is the Dockerfile:
> 
> ```dockerfile
> from ubuntu:18.04
> [...]
> run cd ~/rex/ && dzil authordeps --missing | cpanm
> run cd ~/rex/ && dzil listdeps | cpanm
> ```

Great! As mentioned above the OS-specific dependencies are installed by `dzil install`. That is missing from the Dockerfile, so tests are going to break. If you want to install them manually without running `dzil install`, on a non-Windows system this might work instead of it:

```
cpanm Net::SFTP::Foreign
cpanm IO::Pty
```

There are also a bunch of optional dependencies (see [`.travis.yml`](https://github.com/RexOps/Rex/blob/master/.travis.yml#L27-L29)) you might want to add (and also perhaps MySQL for the database related tests).

You might also be interested in adding `--missing` after `dzil listdeps`, so `cpanm` has to check for less things.

> P.S. distzilla is so damn heavy :)

And also does a heavy-lifting in exchange, and only has to be installed once ;)

**Comment by hurricup** [comment]
Shouldn't contributor be able to run tests without installing Rex? Can't distzilla handle those os-specific dependencies? And again - would be nice to see informative failure :)

**Comment by ferki** [comment]
> Shouldn't contributor be able to run tests without installing Rex?

I already edited the opening post earlier to follow the bug report template, and described the following as "expected behavior":

> The test suite should pass after cloning the repo, and installing the dependencies.

It also already possible by manually installing the dependencies that can't be handled otherwise (e.g. OS-specific dependencies). But this has to be mentioned in the various installation documentation (README, website and contributing guide).

---

> Can't distzilla handle those os-specific dependencies?

I don't think it's possible, but if there's a way, I'm highly interested. As mentioned in my [second comment](https://github.com/RexOps/Rex/issues/1396#issuecomment-671888298):

> On the other hand, it seems that `dzil listdeps` doesn't pick up OS-specific dependencies like `Net::SFTP::Foreign` and `IO::Pty` on Linux (or `Net::SSH2` on Windows). It's because they are injected into `Makefile.PL` through `Dist::Zilla::Plugin::OSPrereqs`, so they are only resolved and installed during `dzil install`.

In other words, OS-specific dependencies are only resolved during install-time. So in that sense it's more of a `Dist::Zilla` issue, and I'm not aware of any way to do it with `dzil listdeps` (fixme).

---

> And again - would be nice to see informative failure :)

Yes, also in the same comment I mentioned:

> As a workaround perhaps we can add an explicit test for OS-specific dependencies in order to have a more useful test failure message, or we could improve the conditions of the tests mentioned here.

**Comment by hurricup** [comment]
Works fine after installing these deps. thanks

**Comment by ferki** [comment]
Thanks again for the report, @hurricup!

I've added some tests to check the presence of OS-specific dependencies, so that should result in a more useful test failure in such cases. I also updated the readme and the contributing guide to include instructions on how to manually install these dependencies if needed. An idea for the future is to reduce or eliminate the need for OS-specific dependencies, but that's a story for another time.

---

Regarding the example Dockerfile you posted earlier: the purpose of both docker and perlbrew is to isolate things from the system. I keep thinking that maybe using only one of them would be simpler. For example there are [official Perl docker images](https://hub.docker.com/_/perl/) so perlbrew could be spared. Or the other way around: perlbrew can handle isolation of the perl environment, and docker could be spared. Based on these learnings, I'm even tempted to create a docker image for Rex.

**Comment by hurricup** [comment]
Thank you!
Regarding docker and brew: system may have different perl by default, and perlbrew is easiest way to fix this. I wanted to reproduce my exact problem with closest environment to my own. That is why it's docker and perlbrew.
